### ISSUE TITLE
xcopy /E also copies folders when copying files

### ISSUE THREAD

**Issue by Meoo** [premake/premake-core#124]
I didn't believe it at first, but xcopy seems to be the worst copy utility ever made.

By default, premake5 adds the "/E" argument which sould mean "recursive, including empty folders".
The problem is when you try to copy a file, it will also copy the hierarchy of the folder this file is in (only folders, no files).
The "/S" (which should mean recursive except empty folders) is also flawed, but it is much more random. Sometimes it copies 1 or 2 folders sometimes it doesn't...
I can't think of any reason it does that, but it does (tested on win7).
To try it just xcopy /E or /S a file which have sibling folders into an empty folder.

It seems the "/E" option should never be used when copying a single file... I blame xcopy on that one.
I have no idea how this can be fixed, robocopy just denies the copy of a single file if /E or /S is used.

**Comment by samsinsane** [comment]
Could this be fixed by changing the copy function defined in os.commandTokens.windows to something like this:
```
copy = function(v)
	v = path.translate(v)
	if os.isdir(v) then
		return "xcopy /Q /E /Y /I " .. v .. " > nul"
	else
		return "xcopy /Q /Y /I " .. v .. " > nul"
	end
end,
```
I'm more than happy to create a PR for this, if this is an acceptable solution to the problem?

**Comment by Meoo** [comment]
Thanks for your input, but in my case (which was rather complex), the file to be copied was the output produced by the build (it didn't exist when running premake), so this wouldn't completely fix the issue.

Really, the problem just comes from xcopy...

**Comment by samsinsane** [comment]
Unless I'm misunderstanding what you're saying, this would work regardless of whether the file existed or not, it has to be a directory to receive the `/E` switch.

**Comment by Meoo** [comment]
Well, in my case yes it would work, but maybe someone will want to copy a folder generated (for example assets baked by another post build command), and in that case the xcopy without /E will not work, while the default behaviour with /E will always work.

To completely fix the issue the file type ought to be checked when running the command, but that would be complex and error prone just to remove some mildly annoying empty folders...

**Comment by samsinsane** [comment]
You're right, I came to this conclusion too after thinking about it a bit more. There's another alternative, the check could be performed when the command is executed. I'm still trying to figure it out, but something like this:
```
return "IF EXIST " .. v .. "\NUL (xcopy /Q /E /Y /I " .. v .. " > nul) ELSE (xcopy /Q /Y /I " .. v .. " > nul)"
```
From what I've been reading it needs to span multiple lines and while this would probably fix the issue, it's not exactly an elegant solution. Thoughts?

**Comment by samsinsane** [comment]
After a bit of playing around, this appears to do the trick:
```
return "IF EXIST " .. v .. "\\NUL (xcopy /Q /E /Y /I " .. v .. " > nul) ELSE (xcopy /Q /Y /I " .. v .. " > nul)"
```
If you're happy with this I can create a PR for it.

**Comment by Meoo** [comment]
I have tried the command manually and the \\NUL trick seems to work fine to detect folders, but be wary that the "v" string may not be formated as you except (it has both the source and the destination, and may contain spaces or quotes for example).

When you have something that works well enough, don't hesitate to send a PR so everyone else can try it and give feedback.

**Comment by tvandijck** [comment]
Has anyone tried using Robocopy instead of xcopy?? it is apparently a better xcopy on windows..
it's been available since XP I believe...

**Comment by samsinsane** [comment]
It would appear the syntax is slightly different, which would require some more changes (and still require the changes in my PR). According to [this website](http://ss64.com/nt/robocopy.html), the syntax for Robocopy is `ROBOCOPY Source_folder Destination_folder [files_to_copy] [options]`, which is a bit different to [xcopy](http://ss64.com/nt/xcopy.html).

I haven't really used either of them that much, but it seems like xcopy fits the situation better. I honestly don't know, I'm happy to defer to someone with experience using them?

**Comment by tritao** [comment]
I've started using `robocopy` for some recursive folder copies in some of my scripts, it's worked well so far.

**Comment by samsinsane** [comment]
@starkos I believe this can be resolved now, since #231 has gone through. Unless more work is required on this? I'm more than happy to look into this again if needed.

**Comment by starkos** [comment]
Thanks for pointing that out…closing now.
